src/v/config/property_constraints.h

```diff
@@ -0,0 +1,268 @@
+  std::optional<ss::sstring> min_str_opt,
+  std::optional<ss::sstring> max_str_opt,
+  constraint_enabled_t enabled);
+
+template<typename T>
+void rjson_serialize_integral(
+  json::Writer<json::StringBuffer>& w, const std::optional<T>& val) {
+    if constexpr (std::is_same_v<T, std::chrono::milliseconds>) {
+        w.Uint64(val->count());
+    } else if constexpr (std::is_same_v<T, uint64_t>) {
+        w.Uint64(*val);
+    } else if constexpr (std::is_same_v<T, uint32_t>) {
+        w.Uint(*val);
+    } else if constexpr (
+      std::is_same_v<T, int32_t> || std::is_same_v<T, int16_t>) {
+        w.Int(*val);
+    } else {
+        static_assert(
+          utils::unsupported_type<T>::value,
+          "Type name is not a valid integral");
+    }
+}
+
+template<typename T>
+struct constraint_validator_range : public constraint_validator_t {
+    using extractor_t =
+      typename ss::noncopyable_function<T(const cluster::topic_configuration&)>;
+    using clamper_t = typename ss::noncopyable_function<void(
+      cluster::topic_configuration&, T&)>;
+
+    std::optional<T> min;
+    std::optional<T> max;
+    extractor_t value_extractor;
+    clamper_t clamper;
+
+    constraint_validator_range(
+      std::optional<T> min,
+      std::optional<T> max,
+      extractor_t value_extractor,
+      clamper_t clamper)
+      : min{std::move(min)}
+      , max{std::move(max)}
+      , value_extractor{std::move(value_extractor)}
+      , clamper{std::move(clamper)} {}
+
+    bool
+    validate(const cluster::topic_configuration& topic_cfg) const override {
+        T topic_val = value_extractor(topic_cfg);
+
+        if (min && *min > topic_val) {
+            return false;
+        }
+
+        if (max && *max < topic_val) {
+            return false;
+        }
+
+        return true;
+    }
+
+    void clamp(cluster::topic_configuration& topic_cfg) const override {
+        T topic_val = value_extractor(topic_cfg);
+        if (min) {
+            topic_val = std::max(*min, topic_val);
+        }
+        if (max) {
+            topic_val = std::min(*max, topic_val);
+        }
+        clamper(topic_cfg, topic_val);
+    }
+
+    void encode_yaml(YAML::Node& n) const override {
+        if (min) {
+            n["min"] = *min;
+        }
+
+        if (max) {
+            n["max"] = *max;
+        }
+    }
+
+    void rjson_serialize(json::Writer<json::StringBuffer>& w) const override {
+        w.Key("min");
+        if (min) {
+            rjson_serialize_integral<T>(w, min);
+        } else {
+            w.Null();
+        }
+
+        w.Key("max");
+        if (max) {
+            rjson_serialize_integral<T>(w, max);
+        } else {
+            w.Null();
+        }
+    }
+
+    void print(std::ostream& os) const override {
+        os << fmt::format("min: {}", min);
+        os << fmt::format("max: {}", max);
+    }
+};
+
+template<typename T>
+struct constraint_validator_enabled : public constraint_validator_t {
```

Comment:
suggestion: I think it would be helpful adding a comment that describes what this validator is doing vs the range one.  "enabled" makes it seem that constraint validation is enabled, not that you're locking the value of the config.